### PR TITLE
feat(metrics): Split buckets into partitions (dry run) [INGEST-1472]

### PR DESCRIPTION
**Pull request by jjbayer**
Measure what distributions we would obtain if we split flush buckets not only by project, but also by partition, which is determined by a configurable number of partitions and hashing the bucket key.

#skip-changelog

### CODE REVIEW

File: relay-metrics/src/aggregation.rs [hunk 1]
```diff
@@ -1717,7 +1733,13 @@ impl Aggregator {
                     cost_tracker.subtract_cost(key.project_key, key.cost());
                     cost_tracker.subtract_cost(key.project_key, value.cost());
                     let bucket = Bucket::from_parts(key.clone(), bucket_interval, value);
-                    buckets.entry(key.project_key).or_default().push(bucket);
+                    buckets
+                        .entry(key.project_key)
+                        .or_default()
+                        .push(HashedBucket {
+                            hashed_key: key.as_integer_lossy(), // TODO: Do we need a more reliable hasher?
```
Comment (jjbayer) [inline relay-metrics/src/aggregation.rs]: Not sure if we can rely on this hasher, this comment seems pretty clear:

https://github.com/getsentry/relay/blob/a9d93903f11112ed4d3222646df84c0c9fe24d74/relay-metrics/src/aggregation.rs#L894-L895

Comment (untitaker) [reply relay-metrics/src/aggregation.rs]: I think I looked into this, didn't find any great, fast crates, and then @jan-auer pointed out that we already have FMV for this.

Considering we're doing sharding of kafka topics by org id at some point, I think it's time well spent to look into a hashing function that is deterministic and portable (can be replicated in Python), and I am not sure if this function here fulfills any of those criteria.

Comment (jjbayer) [reply relay-metrics/src/aggregation.rs]: > we already have FMV for this.

Good point, will replace with FnvHasher.

Comment (untitaker) [reply relay-metrics/src/aggregation.rs]: > FnvHasher

If you have the time, ensure we're picking a hashing function that is easily usable in python in any case. while portability not required for this story, we will need it in future tasks for traffic steering

Comment (jjbayer) [reply relay-metrics/src/aggregation.rs]: There's a Python impl for FNV, and it looks simple enough to write ourselves if necessary:
https://pypi.org/project/fnvhash/
https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function

File: relay-metrics/src/aggregation.rs [hunk 2]
```diff
@@ -890,12 +890,15 @@ impl BucketKey {
     /// interact with this trait.
     ///
     /// [`cadence::ext::ToSetValue`]: https://docs.rs/cadence/*/cadence/ext/trait.ToSetValue.html
-    fn as_integer_lossy(&self) -> i64 {
+    fn as_integer_lossy(&self) -> u32 {
         // XXX: The way this hasher is used may be platform-dependent. If we want to produce the
         // same hash across platforms, the `deterministic_hash` crate may be useful.
+
+        // TODO(jjbayer): Use FnvHasher here
```
Comment (untitaker) [inline relay-metrics/src/aggregation.rs]: as discussed, fnv is fine and easy to impl, let's do it

File: relay-metrics/src/aggregation.rs [hunk 3]
```diff
@@ -1729,11 +1753,57 @@ impl Aggregator {
...
+        let partition_tag = match partition_key {
+            Some(partition_key) => format!("{partition_key}"),
+            None => "none".to_owned(),
+        };
+        let capped_batches: Vec<_> = capped_batches.collect();
+        for (i, batch) in capped_batches.into_iter().enumerate() {
+            relay_statsd::metric!(
+                histogram(MetricHistograms::BucketsPerBatch) = batch.len() as f64,
+                partition_key = partition_tag.as_str(),
+                batch_index = format!("{i}").as_str(),
```
Comment (untitaker) [inline relay-metrics/src/aggregation.rs]: please don't tag this, as the cardinality is theoretically unbound, and instead add a histogram metric counting the number of batches within a partition, i.e. `metric!(histogram(..) = capped_batches.len());`

File: relay-metrics/src/aggregation.rs [hunk 4]
```diff
@@ -1729,11 +1753,57 @@ impl Aggregator {
...
+        buckets: impl IntoIterator<Item = Bucket>,
+        partition_key: Option<u32>,
+        mut process: F,
+    ) where
+        F: FnMut(Vec<Bucket>),
+    {
+        let capped_batches =
+            CappedBucketIter::new(buckets.into_iter(), self.config.max_flush_bytes);
+        let partition_tag = match partition_key {
+            Some(partition_key) => format!("{partition_key}"),
```
Comment (untitaker) [inline relay-metrics/src/aggregation.rs]: I think we prefer to use `.to_string()` instead of `format!("{..}")`

File: relay-statsd/src/lib.rs [hunk 5]
```diff
@@ -162,6 +161,26 @@ pub fn set_client(client: MetricsClient) {
...
+    let test_client = MetricsClient {
+        statsd_client: StatsdClient::from_sink("", sink),
+        default_tags: Default::default(),
+        sample_rate: 1.0,
+    };
+    set_client(test_client);
+
+    f();
+
+    *METRICS_CLIENT.write() = old_client;
```
Comment (untitaker) [inline relay-statsd/src/lib.rs]: I think you might have those changes locally, but this needs to be done per-thread

**Review by untitaker — CHANGES_REQUESTED** [review]

File: relay-metrics/Cargo.toml [hunk 6]
```diff
@@ -20,7 +20,7 @@ relay-system = { path = "../relay-system" }
 serde = { version = "1.0.114", features = ["derive"] }
 serde_json = "1.0.55"
 failure = "0.1.8"
-crc32fast = "1.2.1"
+fnv = "1.0.7"
```
Comment (jjbayer) [inline relay-metrics/Cargo.toml:23]: This dependency was already in Cargo.lock.

File: relay-metrics/src/aggregation.rs [hunk 7]
```diff
@@ -881,21 +882,12 @@ struct BucketKey {
...
-    ///
-    /// [`cadence::ext::ToSetValue`]: https://docs.rs/cadence/*/cadence/ext/trait.ToSetValue.html
-    fn as_integer_lossy(&self) -> i64 {
-        // XXX: The way this hasher is used may be platform-dependent. If we want to produce the
-        // same hash across platforms, the `deterministic_hash` crate may be useful.
-        let mut hasher = crc32fast::Hasher::new();
+    // Create a 64-bit hash of the bucket key using FnvHasher.
+    // This is used for partition key computation and statsd logging.
+    fn hash64(&self) -> u64 {
+        let mut hasher = FnvHasher::default();
```
Comment (jjbayer) [inline relay-metrics/src/aggregation.rs:888]: With `fnv::FnvHasher` we can auto-derive `Hash`, with `hash32`, we cannot.

**Review by untitaker — APPROVED** [review]

**Pull request merged**